mitodl/odl-video-service
# Import public TechTV collections and set video stream source

~~Review/merge PR #446 first~~

#### What are the relevant tickets?
Fixes #451

#### What's this PR do?
Adds a `stream_source` attribute to the `Collection` model and adds additional parameters to the TechTV import script to handle public collections:
  - `noyoutube` parameter:  videos for public collection ids following this parameter will not be uploaded to youtube (stream_source set to `Cloudfront`)
  - `cloudfront` parameter: videos for public collection ids following this parameter will be streamable from cloudfront if youtube is not available (stream_source set to `None`)
  - By default, any public collection id's not included in the above parameters will be streamable from Youtube only (stream_source set to `Youtube`)

Adds/modifies post-save signals to `Collection` and `Video` so that public videos will either be uploaded to or deleted from Youtube depending on the value of the collection's `stream_source`.

#### How should this be manually tested?
- Set `ENABLE_VIDEO_PERMISSIONS=True` in your `.env` file
- Set `DROPBOX_TOKEN` (you'll need a dropbox app) and `DROPBOX_FOLDER` (where the `ODL-Video/TechTV/Captions` dropbox folder is shared, your app will need access to it).
- Install/start MySQL and import the TechTV database dump located at https://www.dropbox.com/home/ODL-Video/TechTV/SQL
   ```
    mysql -u <username> -p techtv <dump_file.sql>
    ```
- Run the command: 
  ```docker-compose run web python manage.py techtvimport --user <mysql_user> --db <mysql_db> --host <mysql_host> --protected 0 --aws --collections 3255```
  - The new collection should have a `stream_source` of `Youtube`, and after videofiles are done being copied from S3 there should be a `YoutubeVideo` object for each `Video` in the collection (this may take a few minutes).

- Run the command: 
  ```docker-compose run web python manage.py techtvimport --user <mysql_user> --db <mysql_db> --host <mysql_host> --protected 0 --aws --cloudfront 3255 --collections 3255```
  - The collection should have a `stream_source` of `None`, and there should still be a `YoutubeVideo` object for each `Video` in the collection

- Run the command: 
  ```docker-compose run web python manage.py techtvimport --user <mysql_user> --db <mysql_db> --host <mysql_host> --protected 0 --aws --noyoutube 3255 --collections 3255```
  - The collection should have a `stream_source` of `Cloudfront`, and there should no longer be any `YoutubeVideo` objects for the collection's videos.

- In the admin console, change the collection's `stream_source` to `None` or `Youtube`.  After a minute or two there should be a `YoutubeVideo` object for each `Video` in the collection

- Every video in the collection should have its permission set to `Public`, regardless of which command above is run.

#### Any background context you want to provide?
For #452, the following logic will be used to play public videos:
  - video collection `stream_source` = `None`: play from Youtube if possible, otherwise play from Cloudfront
  - video collection `stream_source` = `Cloudfront`: play from Cloudfront, never Youtube
  - video collection `stream_source` = `Youtube`: play from Youtube, never Cloudfront
